commit 6bff17d0a701518aaa902667474a3289ec3e1f0a
Author: Jeff Law <jeffrey.law@oss.qualcomm.com>
Date:   Mon Aug 31 15:06:02 2026 -0600

    Revert "tree-optimization/126998 - wrong-code with lane-reducing reduction"
    
    This reverts commit d8f1bd8c34a289de0810178d897cb246ce2033cf.

diff --git a/gcc/testsuite/gcc.dg/vect/vect-reduc-dot-pr126998-2.c b/gcc/testsuite/gcc.dg/vect/vect-reduc-dot-pr126998-2.c
deleted file mode 100644
index 66f018c291d2..000000000000
--- a/gcc/testsuite/gcc.dg/vect/vect-reduc-dot-pr126998-2.c
+++ /dev/null
@@ -1,37 +0,0 @@
-/* { dg-additional-options "-O3" } */
-/* { dg-additional-options "-msse4" { target sse4 } } */
-
-#include "tree-vect.h"
-
-unsigned a, c;
-int b, d;
-signed char e;
-  
-void __attribute__((noipa))
-f (void)
-{ 
-  for (; b; ++b)
-    {
-      c = 0;
-      for (; c < 10; c++)
-        {
-          d = 0;
-          for (; d < c; d++)
-            a += e;
-        }
-    }
-}
-
-int
-main ()
-{
-  check_vect ();
-
-  b = -3;
-  e = 2;
-  f ();
-  if (a != 3 * 45 * 2)
-    __builtin_abort ();
-
-  return 0;
-}
diff --git a/gcc/tree-vect-loop.cc b/gcc/tree-vect-loop.cc
index cb961440656d..78f84c6269a5 100644
--- a/gcc/tree-vect-loop.cc
+++ b/gcc/tree-vect-loop.cc
@@ -6854,10 +6854,9 @@ vectorizable_lane_reducing (loop_vec_info loop_vinfo, stmt_vec_info stmt_info,
 
   /* Compute number of effective vector statements for costing from the
      number of input lanes allow for excess lanes in the last input vector.  */
-  unsigned int ncopies_for_cost, excess_elts;
+  unsigned int ncopies_for_cost;
   bool res = vect_get_num_copies_for_invariant (loop_vinfo, node_in,
-						&ncopies_for_cost,
-						&excess_elts);
+						&ncopies_for_cost);
   gcc_assert (res && ncopies_for_cost >= 1);
 
   if (vect_is_emulated_mixed_dot_prod (slp_node))
diff --git a/gcc/tree-vect-slp.cc b/gcc/tree-vect-slp.cc
index 06db2f288367..bcff0fb4ec6c 100644
--- a/gcc/tree-vect-slp.cc
+++ b/gcc/tree-vect-slp.cc
@@ -8925,14 +8925,12 @@ vect_scalar_ops_slice_hash::equal (const value_type &s1,
 }
 
 /* Like vect_get_num_copies but N copies of the vector might have
-   excess elements in the last vector.  Stores the number of excess
-   elements in the last vector in *EXCESS_ELTS.
-   Returns false if *NVECTORS or *EXCESS_ELTS cannot be computed.  */
+   excess elements in the last vector.  Returns false if *NVECTORS
+   cannot be computed.  */
 
 bool
 vect_get_num_copies_for_invariant (vec_info *vinfo, slp_tree node,
-				   unsigned *nvectors,
-				   unsigned *excess_elts)
+				   unsigned *nvectors)
 {
   poly_uint64 vf;
 
@@ -8943,12 +8941,8 @@ vect_get_num_copies_for_invariant (vec_info *vinfo, slp_tree node,
   vf *= SLP_TREE_LANES (node);
 
   tree vectype = SLP_TREE_VECTYPE (node);
-  uint64_t rem;
-  bool res = (can_div_away_from_zero_p (vf, TYPE_VECTOR_SUBPARTS (vectype),
-					nvectors)
-	      && ((TYPE_VECTOR_SUBPARTS (vectype) * *nvectors - vf)
-		  .is_constant (&rem)));
-  *excess_elts = rem;
+  bool res = can_div_away_from_zero_p (vf, TYPE_VECTOR_SUBPARTS (vectype),
+				       nvectors);
   return res;
 }
 
@@ -9132,9 +9126,8 @@ vect_slp_analyze_node_operations (vec_info *vinfo, slp_tree node,
 	    }
 
 	  /* Make sure we can generate them and then cost them.  */
-	  unsigned nvectors, excess_elts;
-	  if (!vect_get_num_copies_for_invariant (vinfo, node, &nvectors,
-						  &excess_elts))
+	  unsigned nvectors;
+	  if (!vect_get_num_copies_for_invariant (vinfo, node, &nvectors))
 	    return false;
 	  vect_prologue_cost_for_slp (child, nvectors, cost_vec);
 	}
@@ -10903,10 +10896,9 @@ vect_create_constant_vectors (vec_info *vinfo, slp_tree op_node)
   /* We always want SLP_TREE_VECTYPE (op_node) here correctly set.  */
   vector_type = SLP_TREE_VECTYPE (op_node);
 
-  unsigned int number_of_vectors, excess_elts;
+  unsigned int number_of_vectors;
   bool res = vect_get_num_copies_for_invariant (vinfo, op_node,
-						&number_of_vectors,
-						&excess_elts);
+						&number_of_vectors);
   gcc_assert (res);
   SLP_TREE_VEC_DEFS (op_node).create (number_of_vectors);
   auto_vec<tree> voprnds (number_of_vectors);
@@ -10939,7 +10931,8 @@ vect_create_constant_vectors (vec_info *vinfo, slp_tree op_node)
   tree_vector_builder elts (vector_type, nunits, 1);
   elts.quick_grow (nunits);
   /* Zero-pad the last vector if necessary.  */
-  number_of_places_left_in_vector = nunits - excess_elts;
+  number_of_places_left_in_vector
+    = nunits - nunits * number_of_vectors % group_size;
   for (i = nunits; i > number_of_places_left_in_vector; --i)
     elts[i-1] = build_zero_cst (TREE_TYPE (vector_type));
   stmt_vec_info insert_after = NULL;
diff --git a/gcc/tree-vectorizer.h b/gcc/tree-vectorizer.h
index b3fe5acb60a9..4d542d57e6df 100644
--- a/gcc/tree-vectorizer.h
+++ b/gcc/tree-vectorizer.h
@@ -2790,7 +2790,7 @@ extern tree prepare_vec_mask (loop_vec_info, tree, tree, tree,
 extern tree vect_get_mask_load_else (int, tree);
 extern bool vect_load_perm_consecutive_p (slp_tree, unsigned = UINT_MAX);
 extern bool vect_get_num_copies_for_invariant (vec_info *, slp_tree,
-					       unsigned *, unsigned *);
+					       unsigned *);
 
 /* In tree-vect-patterns.cc.  */
 extern void
